### PR TITLE
Fixed duplicating endpoints

### PR DESCRIPTION
**Pull request by u2**
#552

### CODE REVIEW

File: spec/grape/dsl/routing_spec.rb [hunk 1]
```diff
@@ -82,9 +82,17 @@ class Dummy
...
-        it 'defines a new endpoint' do
-          expect { subject.route(:any) }
-            .to change{ subject.endpoints.count }.from(0).to(1)
+        context 'defines new endpoint' do
+          it 'defines a new endpoint' do
+            expect { subject.route(:any) }
+              .to change{ subject.endpoints.count }.from(0).to(1)
+          end
+
+          it 'uniq the duplicating endpoint' do
```
Comment (dblock) [inline spec/grape/dsl/routing_spec.rb]: You don't need to put these two tests together, just add a new test at the same level that says 'does not duplicate identical endpoints'.

**Comment by dblock** [comment]
I understand the case of a reloader, but is this generally correct? What happens when you re-define an endpoint with all the same arguments, you probably would want the second one, not the first one, to be defined or to see a warning/error?

Maybe a better approach could be to freeze endpoints or something like that? Not sure if it can be accomplished.

Definitely needs a CHANGELOG, thanks.

File: lib/grape/dsl/routing.rb [hunk 2]
```diff
@@ -120,7 +120,8 @@ def route(methods, paths = ['/'], route_options = {}, &block)
             }).deep_merge(route_setting(:description) || {}).deep_merge(route_options || {})
           }
 
-          endpoints << Grape::Endpoint.new(inheritable_setting, endpoint_options, &block)
+          new_endpoint = Grape::Endpoint.new(inheritable_setting, endpoint_options, &block)
+          endpoints << new_endpoint unless endpoints.any?{ |o| o.options == new_endpoint.options && o.inheritable_setting.to_hash == new_endpoint.inheritable_setting.to_hash }
```
Comment (dblock) [inline lib/grape/dsl/routing.rb]: I would want a comparison operator on `Grape::Endpoint` so we could maybe write `endpoints.include?(...)` or `endpoints.any? { |e| endpoint.equals?(e) } # or == `.

Comment (u2) [inline lib/grape/dsl/routing.rb]: Yes, The the comparison of `Endpoint` is meaningful.

**Comment by u2** [comment]
Yes.I can't agree more.Before, I did not think so much about that.

**Comment by dblock** [comment]
Merging.
